fdbserver/SimEncryptVaultProxy.actor.cpp

```diff
@@ -0,0 +1,153 @@
+				SimGetEncryptKeyByKeyIdReply reply;
+
+				// Lookup corresponding EncryptKeyCtx for input keyId
+				if (vaultProxyCtx.simEncryptKeyStore.find(req.encryptKeyId) != vaultProxyCtx.simEncryptKeyStore.end()) {
+					reply.encryptKey = StringRef(vaultProxyCtx.simEncryptKeyStore[req.encryptKeyId].get()->key);
+					req.reply.send(reply);
+				} else {
+					req.reply.sendError(key_not_found());
+				}
+			}
+			when(SimGetEncryptKeyByDomainIdRequest req = waitNext(interf.encryptKeyLookupByDomainId.getFuture())) {
+				SimGetEncryptKeyByDomainIdReply reply;
+
+				// Map encryptionDomainId to corresponding EncryptKeyCtx element using a modulo operation. This would
+				// mean multiple domains gets mapped to the same encryption key which is fine, the EncryptKeyStore
+				// guarantees that keyId -> plaintext encryptKey mapping is idempotent.
+
+				reply.encryptKeyId = req.encryptDomainId % maxEncryptKeys;
+				reply.encryptKey = StringRef(vaultProxyCtx.simEncryptKeyStore[reply.encryptKeyId].get()->key);
+				req.reply.send(reply);
+			}
+		}
+	}
+}
+
+void forceLinkSimEncryptVaultProxyTests() {}
+
+namespace {
+
+ACTOR Future<Void> testRunWorkload(SimEncryptVaultProxyInterface inf, uint32_t nEncryptionKeys) {
+	state uint32_t maxEncryptionKeys = nEncryptionKeys;
+	state int maxDomainIds = deterministicRandom()->randomInt(121, 295);
+	state int maxIterations = deterministicRandom()->randomInt(786, 1786);
+	state std::unordered_map<SimEncryptDomainId, std::unique_ptr<SimEncryptKeyCtx>> domainIdKeyMap;
+	state int i = 0;
+
+	TraceEvent("RunWorkloadStart").detail("MaxDomainIds", maxDomainIds).detail("MaxIterations", maxIterations);
+
+	{
+		// construct domainId to EncryptKeyCtx map
+		for (i = 0; i < maxDomainIds; i++) {
+			SimGetEncryptKeyByDomainIdRequest req;
+			req.encryptDomainId = i;
+			SimGetEncryptKeyByDomainIdReply reply = wait(inf.encryptKeyLookupByDomainId.getReply(req));
+			domainIdKeyMap[i] =
+			    std::make_unique<SimEncryptKeyCtx>(reply.encryptKeyId, reply.encryptKey.toString().c_str());
+		}
+
+		// randomly pick any domainId and validate if lookupByKeyId result matches
+		for (i = 0; i < maxIterations; i++) {
+			state int idx = deterministicRandom()->randomInt(0, maxDomainIds);
+			state SimEncryptKeyCtx* ctx = domainIdKeyMap[idx].get();
+			SimGetEncryptKeyByKeyIdRequest req(ctx->id);
+			SimGetEncryptKeyByKeyIdReply reply = wait(inf.encryptKeyLookupByKeyId.getReply(req));
+			ASSERT(reply.encryptKey.compare(ctx->key) == 0);
+		}
+	}
+
+	{
+		// Verify unknown key access returns the error
+		state SimGetEncryptKeyByKeyIdRequest req;
+		req.encryptKeyId = maxEncryptionKeys + 1;
+		try {
+			SimGetEncryptKeyByKeyIdReply reply = wait(inf.encryptKeyLookupByKeyId.getReply(req));
+		} catch (Error& e) {
+			ASSERT(e.code() == error_code_key_not_found);
+		}
+	}
+
+	TraceEvent("RunWorkloadDone").log();
+	return Void();
+}
+
+} // namespace
+
+TEST_CASE("fdbserver/SimEncryptVaultProxy") {
```

Comment:
Just to clarify, the reason for this test is to exercise the simulated key vault logic that will later be integrated with the rest of the project to provide this functionality in simulation, but since those parts of the project are not complete yet, we want to validate the basic functionality of the simulated key vault?

Reply:
Spot on! Yes, I would update EncyrptKeyServer to leverage SimVaultProxy in simulation as there won't be any real external KeyManagement server listening for those blocks. Also, didn't update EncryptKeyServer as good to get other outstanding review, otherwise it may add to rebasing pains later :).